google/cloud/spanner/internal/session_pool.h

```diff
@@ -0,0 +1,67 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_SESSION_POOL_H_
+#define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_SESSION_POOL_H_
+
+#include "google/cloud/spanner/internal/session.h"
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <memory>
+#include <mutex>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+class SessionManager;
+
+/**
+ * Maintains a pool of `Session` objects.
+ */
+class SessionPool {
+ public:
+  /**
+   * Create a `SessionPool`. Uses `manager` to create, delete, and refresh
+   * sessions in the pool.
+   */
+  SessionPool(SessionManager* manager) : manager_(manager) {}
+
+  /**
+   * Allocate a session from the pool, creating a new `Session` if necessary.
+   * @returns an error if session creation fails; always returns a non-null
+   * pointer on success.
+   */
+  StatusOr<std::unique_ptr<Session>> Allocate();
+
+  /// Release `session` back to the pool.
+  void Release(Session* session);
```

Comment:
Why the asymmetry on `unique_ptr` -v- raw pointer between `Allocate()` and `Release()`.